src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/LanguageServices/EditorConfigFileWatcher.cs

```diff
@@ -0,0 +1,271 @@
+                    ITargetBlock<IProjectVersionedValue<Tuple<IProjectTreeSnapshot, IProjectSubscriptionUpdate>>> target = DataflowBlockSlim.CreateActionBlock<IProjectVersionedValue<Tuple<IProjectTreeSnapshot, IProjectSubscriptionUpdate>>>(DataFlow_ChangedAsync);
+
+                    // Join the two sources so that we get synchronized versions of the data.
+                    _treeWatcher = ProjectDataSources.SyncLinkTo(treeSource, editorconfigItemsSource, target);
+
+                    return Task.CompletedTask;
+                });
+        }
+
+        private CancellationTokenSource CreateLinkedCancellationTokenSource()
+        {
+            // we want to cancel when we switch what file is watched, or when the project is unloaded
+            if (_projectServices.Project?.Services?.ProjectAsynchronousTasks?.UnloadCancellationToken != null)
+            {
+                return CancellationTokenSource.CreateLinkedTokenSource(
+                    _projectServices.Project.Services.ProjectAsynchronousTasks.UnloadCancellationToken);
+            }
+            else
+            {
+                return new CancellationTokenSource();
+            }
+        }
+
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            if (initialized)
+            {
+                _taskDelaySchedulerCancellationTokenSource.Cancel();
+                _taskDelaySchedulerCancellationTokenSource.Dispose();
+                _taskDelayScheduler.Dispose();
+                _treeWatcher.Dispose();
+
+                var filesBeingWatched = _filesBeingWatched.Keys.ToArray();
+                await UnregisterFileWatchersAsync(filesBeingWatched);
+            }
+        }
+
+        private static SortedSet<string> GetEditorConfigFilePaths(IProjectTree newTree, IProjectSubscriptionUpdate projectUpdate)
+        {
+            var editorConfigFilePaths = new SortedSet<string>(StringComparer.OrdinalIgnoreCase);
+            string projectFilePath = projectUpdate.CurrentState.GetPropertyOrDefault(ConfigurationGeneral.SchemaName, ConfigurationGeneral.MSBuildProjectFullPathProperty, defaultValue: null);
+            if (string.IsNullOrEmpty(projectFilePath))
+            {
+                return editorConfigFilePaths;
+            }
+
+            // TODO (tmeschter): if $(DiscoverEditorConfigFiles) is "false", disable all file watchers by returning an empty set.
+
+            string projectDirectory = Path.GetDirectoryName(projectFilePath);
+            string directory = Path.GetDirectoryName(projectDirectory);
+            while (directory != null)
+            {
+                string editorConfigFilePath = Path.Combine(directory, ".editorconfig");
+                editorConfigFilePaths.Add(editorConfigFilePath);
+            }
```

Comment:
Needs `directory = Path.GetDirectoryName(projectDirectory);` in here to prevent infinite loop :P